Responsive layout

johnpryan opened the pull request:
This uses a LayoutBuilder to switch between layouts when the screen width is too small (at about 720px the app bar and footer start to throw layout errors)

Pixel 7 dimensions:

<img width="396" alt="Example of UI on Pixel 7" src="https://github.com/dart-lang/dart-pad/assets/1145719/e0fa3657-bfac-4f62-9b9e-246c5e900dc4">

Fixes https://github.com/dart-lang/dart-pad/issues/2751

johnpryan commented:
There's a bug where changing the layout causes the the CodeMirror editor to appear blank until it's focused. I'm not sure why that's happening yet.

parlough commented:
I haven't tried this out yet, but looking at the screenshot, I'm quite excited!

> There's a bug where changing the layout causes the the CodeMirror editor to appear blank until it's focused. I'm not sure why that's happening yet.

I see the editor go blank all the time, haven't determined a consistent reason yet.

---

Quick note from the screenshot, I'd drop the title in this layout.

parlough commented:
I haven't reviewed the code yet, but I did try running it locally, and this is amazing @johnpryan!! I think this will make DartPad actually useable on mobile, at the very least for reading and trying out samples. Developers will definitely appreciate it.

A few non-blocking issues and suggestions:

- Sometimes the first run fails. Not always? I'll try to reproduce more consistently tomorrow.
- I think it's better for the tabs to not have text like you have, but can you make sure they have accurate semantic labels: "Code", "UI", "Output" or similar?
- Thanks for getting rid of the title, much cleaner! With the extra space, can we add the New and Sample buttons back? 
  - I was thinking next to the brightness button with just their icons and semantic/hover labels.
  - If it feels too cramped, I'd choose to just the sample one. Without those, there's no clear way to start over or explore the available samples.
- Add access to SDK and package information popup from overflow menu.

Some ideas for follow-up:

- As we discussed, adding badges when there's new UI or console output.
  - Or does it make sense to switch automatically once compiled? To the UI tab if a Flutter app is running successfully, otherwise to the console for Dart output or compiler output.

devoncarew commented:
> Or does it make sense to switch automatically once compiled? To the UI tab if a Flutter app is running successfully, otherwise to the console for Dart output or compiler output.

+1, I think auto-switching (when execution starts) makes sense.

qq - do we show compilation status anywhere (overlayed over the bottom of the code?)?

devoncarew commented on pkgs/sketch_pad/lib/main.dart:
```diff
@@ -33,6 +33,7 @@ import 'widgets.dart';
 // TODO: implement find / find next
 
 const appName = 'DartPad';
+const smallScreenWidth = 720;
```
👍 - seems like a good size to pivot at

devoncarew commented on pkgs/sketch_pad/lib/main.dart:
```diff
@@ -217,16 +218,28 @@ class DartPadMainPage extends StatefulWidget {
 
+    tabController = TabController(length: 3, vsync: this)
+      ..addListener(() {
+        setState(() {});
```
Can you add a comment here? This causes a rebuild on tab selected index changes?

devoncarew commented on pkgs/sketch_pad/lib/main.dart:
```diff
@@ -271,215 +284,129 @@ class _DartPadMainPageState extends State<DartPadMainPage> {
+      appModel: appModel,
+      key: _executionWidgetKey,
+    );
+    final loadingOverlay = LoadingOverlay(
```
+1 - great refactoring (here and `EditorWithButtons`); it helps us reason about the responsive layout.

devoncarew commented on pkgs/sketch_pad/lib/main.dart:
```diff
@@ -271,215 +284,129 @@ class _DartPadMainPageState extends State<DartPadMainPage> {
+        );
+    final scaffold =
+        LayoutBuilder(builder: (context, BoxConstraints constraints) {
+      if (constraints.maxWidth <= smallScreenWidth) {
```
`// Build the mobile UI for small screen widths.`

devoncarew commented on pkgs/sketch_pad/lib/main.dart:
```diff
@@ -271,215 +284,129 @@ class _DartPadMainPageState extends State<DartPadMainPage> {
+        );
+      }
+
+      return Scaffold(
```
`// Return the desktop screen UI.`

devoncarew commented on pkgs/sketch_pad/lib/main.dart:
```diff
@@ -566,6 +493,210 @@ class _DartPadMainPageState extends State<DartPadMainPage> {
+                  padding: const EdgeInsets.all(denseSpacing),
+                  child: Row(
+                    mainAxisAlignment: MainAxisAlignment.end,
+                    // We use explicit directionality here in
```
nit: we can re-flow this line comment now

devoncarew reviewed (approved):
Like w/ @parlough's comments, we may want to iterate on this (how much should we show in the toolbar? should we push some toolbar / status line actions into the overflow menu for mobile? add a hamburger menu?). I'm happy to figure that out in this PR or later ones; as is, this is a big improvement for smaller screens.

johnpryan commented:
I think I've addressed most of the feedback here, so I'm going to go ahead and merge so we can keep iterating. Let's make sure we're filing issues for the UX improvements we're missing.
